[Zhang Weiqiang] iP

wweqg opened the pull request:
# Duke

> **An organised and efficient task manager that can be customised to your own taste.**

Duke is a flexible software that helps to plan, manage and track all your tasks. It has,

+ Command line interface
+ Simple commands
+ ~~Smooth~~ *Super Smooth* user experience 

**And it is **<u>FREE!!!</u>**** 

All you need to do is, 

1. download it from [here](https://github.com/wweqg/ip/releases/download/v0.6/duke.jar).
2. place it in the directory that you would like to place your task list at.
3. open your terminal in the same directory, enter `java -jar duke.jar`.
4. follow this [tutorial](#tutorial) to add your tasks. :blush:

**Features**: 

- [x] Manage task
- [ ] Reminders (coming soon)
- [ ] Customisation (coming soon)

----

## Tutorial

Here is a quick tutorial on the command syntax !!! :+1:

Content inside the { } is to be decided by the user. 

- `todo {Task}` To add a Todo task.
- `deadline {Task} \by {'Date in' YYYY-MM-DD}` To add a deadline task.
- `event {Task} \at {'Date in' YYYY-MM-DD}` To add an event task.
- `mark {task index}` To mark the task at specific index in the list as done.
- `unmark {task index}` To mark the task at specific index in the list as not done.
- `delete {task index}` To delete the task at specific index.
- `list` To check the current list of task.
- `help` To check the list of commands.
- `bye` To exit the programme.

Commit:
Add Gradle support

wweqg committed:
Implemented a skeletal version of Duke that starts by greeting the user,
simply echos commands entered by the user, and exits when the user types bye.

wweqg committed:
Level 2: used arraylist to store the entered text.

wweqg committed:
Add the ability to mark tasks as done and the ability to change the status back to not done.
Include a task class to encapsulate the information.

wweqg committed:
Add support for tracking three types of tasks: ToDos, Events, Deadlines.
Reform the structure of the code.

wweqg committed:
Test using the I/O redirection technique

Use the input/output redirection technique to semi-automate the testing of Duke.

wweqg committed:
Teach Duke to deal with some errors such as incorrect inputs entered by the user.
Abstract the available commands to an Enum class.

wweqg committed:
Included delete feature.
Include exception classes.

wweqg committed:
Wrap the whole project with Model, View, Controller structure.
Teach Duke to deal with some more errors.
Fixed some minor bugs.

wweqg committed:
Adjust formatting

wweqg committed:
Divide classes into packages.

Previously the classes are clustered under the java folder, which makes
folder looked unorganized. After splitting them into different packages
(bases on the functionality of each class), the organization of project
is easier.

wweqg committed:
Iclude save feature in the taskList.

Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when Duke starts up

wweqg committed:
Teach Duke how to understand dates and times.

wweqg committed:
Merge branch 'branch-Level-7'

wweqg committed:
Merge branch 'branch-Level-8'

wweqg committed:
Refactor the code to extract out closely related code as classes.

wweqg committed:
Modify features in HELP command.

wweqg committed:
Merge remote-tracking branch 'origin/add-gradle-support'

wweqg committed:
Fixed some errors

wweqg committed:
Add JUnit tests to test the behavior of the parser.

wweqg committed:
Enabled checkstyle plugin in the gradle

wweqg committed:
Fixed some bugs.

wweqg committed:
Fixed some bugs.

wweqg committed:
Give users a way to find a task by searching for a keyword.

wweqg committed:
Modified UI setting.

wweqg committed:
Added JavaDoc comments to the code.

wweqg committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Added JavaDoc comments to the code.

# Conflicts:
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/storage/Storage.java

wweqg committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Modified UI setting.
  Give users a way to find a task by searching for a keyword.

# Conflicts:
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/ui/Ui.java

wweqg committed:
Fixed the conflicts

wweqg committed:
Fixed bugs

wweqg committed:
Adjusted some alignments

wweqg committed:
Adjusted more allignments

wweqg committed:
Deleted unnecessary files.

wweqg committed:
Fixed some bugs.
Included checkstyle plugin.
Used Varargs.

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Added GUI for Duke.

wweqg committed:
Added GUI for Duke.

wweqg committed:
Merge branch 'Level-10'

* Level-10:
  Added GUI for Duke.

wweqg committed:
Fixed checkstyle issue

wweqg committed:
Update README.md

linuschancs commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * The entry of the Duke application
+     *
+     * @param args user input.
```
Can consider capitalizing 'user' as it is the start of the sentence.
```suggestion
     * @param args User input.
```

linuschancs commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+import duke.ui.Ui;
+
+/**
+ * A Duke class that encapsulates the information of Duke
```
Remember to end your class header comments, method header comments with a punctuation!

linuschancs commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+    private final Ui ui;
+
+    /**
+     * Constructs a Duke object
```
Punctuation at the end of method header.

wweqg committed:
Fixed bugs.

wweqg committed:
Merge branch 'master' of https://github.com/wweqg/ip

wweqg committed:
Fixed some bugs.

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Update README.md

kavantan commented on README.md:
```diff
@@ -1,24 +1,42 @@
+
+----
+
+## Tutorial
+
+Here is a quick tutorial on the command syntax !!! :+1:
+
+Content inside the { } is to be decided by the user. 
+
+- `todo {Task}` To add a Todo task.
+- `deadline {Task} \by {'Date in' YYYY-MM-DD}` To add a deadline task.
+- `event {Task} \at {'Date in' YYYY-MM-DD}` To add an event task.
+- `mark {task index}` To mark the task at specific index in the list as done.
+- `unmark {task index}` To mark the task at specific index in the list as not done.
+- `delete {task index}` To delete the task at specific index.
+- `list` To check the current list of task.
+- `help` To check the list of commands.
+- `bye` To exit the programme.
```
I like how you added a tutorial on your DukeBot's usage 👍

kavantan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,53 @@
+     * @param input User input
+     * @return Response from Duke
+     */
+    public String getResponse(String input) {
```
It's great how you followed the coding convention for your methods!

kavantan commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,31 @@
+/**
+ * A command class that encapsulates the action of adding a specific task into the task list.
+ */
+public class AddCommand extends Command {
```
It's great that you followed the coding convention for your class names as well!

linuschancs replied:
Indeed. Good naming convention used for variable names, class names!

kavantan commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,45 @@
+package duke.command;
+
+import duke.exception.FileIoException;
+import duke.exception.InvalidIndexException;
+import duke.storage.Storage;
+import duke.task.Task;
+import duke.task.TaskList;
+import duke.ui.Ui;
```
It's great how you followed the coding convention for your imports as well!

kavantan commented on src/main/java/duke/exception/InvalidIndexException.java:
```diff
@@ -0,0 +1,10 @@
+package duke.exception;
+
+/**
+ * An Exception class that encapsulates the situation where the user has given an invalid index to a task.
+ */
```
I like how meticulous you are with your JavaDocs, writing clear documentation for all of your Exception classes 👍

kavantan commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,179 @@
+    public static Command parse(String fullCommand) throws DukeException {
+        String[] arr = fullCommand.trim().split("\\s+", 2);
+        String taskType = arr[TASK_TYPE];
+        switch (taskType.toUpperCase()) {
+        case "BYE":
+            return new ExitCommand();
```
It's fantastic that your switch & case statements adhere to the coding standard too!

kavantan reviewed (commented):
Overall, your code was very well written! As far as I can tell, your codebase follows all the coding conventions very meticulously 👍

linuschancs commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,179 @@
+ */
+public class Parser {
+
+    private static final int TASK_TYPE = 0;
+    private static final int COMMAND_DESCRIPTION = 1;
```
Good naming convention used here for constants!

linuschancs reviewed (commented):
Overall, I really liked how you abstracted out your classes to fulfil multiple functionalities, this really shows the amount of thought you put into this.
Most of the code also follows very good coding conventions, the only main improvement would probably be to be more careful with punctuations for header comments and starting the JavaDocs sentence with a capital letter!

wweqg committed:
no message

wweqg committed:
Set up Continuous Integration (CI) for the project

This project has no CI set up

When merging PR, no tests were carry out

Add a gradle.yml file to include CI support

Follows the ip instruction.

wweqg committed:
Use assert feature

The program had no assert feature

The input and parameter were assumed to be correct as there were
no extra checks

Include assert feature

By using the built in assert method in Java

wweqg committed:
Merge pull request #1 from wweqg/A-Assertions

Use assert feature

wweqg committed:
Improve code quality

Some of the codes in the same method does not have the same level of
abstraction.
Also, redundant and unclear codes are present.

Removed the unnecessary codes and refactor the other code to
ensure the same level of abstraction.

wweqg committed:
Merge pull request #2 from wweqg/A-CodeQuality

Improve code quality

wweqg committed:
Support stream operations

The program had no use of stream.

Some of the possible extensions could be ineffective without using
stream operations.

Include a method to get a stream representation of all the tasks.

This eases the process of building possible extension in the future.

wweqg committed:
Merge pull request #3 from wweqg/A-Streams

Support stream operations

wweqg committed:
Merge branch 'master' of https://github.com/wweqg/ip

* 'master' of https://github.com/wweqg/ip:
  Support stream operations

wweqg committed:
Add reschedule feature

The added task cannot be reschedule/modified.

Users had to delete the task then add a new task. However, this
approach will alter the sequence of the tasks in the list.

Include a reschedule command

format: reschedule {index} {Task}
This allows user to modify the saved task without changing the
sequence.

wweqg committed:
Refine code further

Many classes lack of SLAP

Parser class was in a mess

improved the code in these classes, renamed resechdule command to edit
command.

wweqg committed:
Update javadocs

wweqg committed:
Include background color for GUI

wweqg committed:
Add files via upload

wweqg committed:
Delete Screenshot 2022-09-17 191502.png

wweqg committed:
Add files via upload

wweqg committed:
Delete duke_demo.png

wweqg committed:
Add files via upload

wweqg committed:
Delete Ui.png

wweqg committed:
Add files via upload

wweqg committed:
no message

wweqg committed:
Merge branch 'master' of https://github.com/wweqg/ip

wweqg committed:
Create _config.yml

wweqg committed:
Update _config.yml

wweqg committed:
Update _config.yml

wweqg committed:
Update _config.yml

wweqg committed:
Update _config.yml

wweqg committed:
Update _config.yml

wweqg committed:
Update _config.yml

wweqg committed:
no message

wweqg committed:
Delete README.md

wweqg committed:
Add files via upload

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Update _config.yml

wweqg committed:
Update _config.yml

wweqg committed:
Update README.md

wweqg committed:
no message

wweqg committed:
Update README.md

wweqg committed:
Update README.md

wweqg committed:
Update README.md
